percona_pg_telemetry.c

```diff
@@ -780,12 +826,153 @@ get_extensions_list(PTDatabaseInfo *dbinfo, MemoryContext cxt)
 	return extlist;
 }
 
+static List *
+get_access_methods_list(MemoryContext cxt)
+{
+	List	   *amlist = NIL;
+	SPITupleTable *tuptable;
+	int			spi_result;
+	char	   *query = "SELECT am.amname AS access_method, count(*) FROM pg_class c LEFT JOIN pg_am am ON c.relam = am.oid WHERE c.relkind = 'r' AND c.relnamespace NOT IN ( SELECT oid FROM pg_namespace WHERE nspname IN ('pg_catalog', 'information_schema', 'pg_toast') ) GROUP BY access_method ORDER BY access_method;";
+	MemoryContext oldcxt;
+
+	SetCurrentStatementStartTimestamp();
+	StartTransactionCommand();
+
+	/* Initialize SPI */
+	if (SPI_connect() != SPI_OK_CONNECT)
+	{
+		ereport(ERROR, (errmsg("Failed to connect to SPI")));
+	}
+
+	PushActiveSnapshot(GetTransactionSnapshot());
+
+	/* Execute the query */
+	spi_result = SPI_execute(query, true, 0);
+	if (spi_result != SPI_OK_SELECT)
+	{
+		SPI_finish();
+		ereport(ERROR, (errmsg("Failed to execute query")));
+	}
+	if (SPI_processed > 0)
+	{
+		PTAccessMethodInfo *aminfo;
+		uint64		row_count;
+
+		char	   *amname;
+		char	   *count;
+
+		/* Switch to our memory context instead of the transaction one */
+		oldcxt = MemoryContextSwitchTo(cxt);
+
+		tuptable = SPI_tuptable;
+		for (row_count = 0; row_count < SPI_processed; row_count++)
+		{
+			HeapTuple	tuple = tuptable->vals[row_count];
+
+			aminfo = (PTAccessMethodInfo *) palloc(sizeof(PTAccessMethodInfo));
```

Comment:
I wonder if we should not get into the habit of using `palloc_object()` for things like this. It was backported to PostgreSQL 10 so it should be safe to use.